type-metadata/src/lib.rs

```diff
@@ -37,12 +37,21 @@ pub enum IdentKind {
 	Array(ArrayIdent),
 	Slice(SliceIdent),
 	Tuple(TupleIdent),
-	Option(OptionIdent),
-	Result(ResultIdent),
 }
 
 #[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
-pub struct Namespace(Vec<&'static str>);
+pub struct Namespace {
+	segments: Vec<&'static str>,
+}
+impl Namespace {
+	pub fn new(segments: Vec<&'static str>) -> Self {
```

Comment:
Hide implementation details with:
```rust
pub fn new<S>(segments: S) -> Self
where
    S: IntoIterator<Item = &'static str>,
{
    Namespace {
        segments: segments.into_iter().collect(),
    }
}
```
This will result in the same code for `Vec` but allows for other inputs and hide the underlying implementation.

Reply:
Maybe with the existence of `prelude` we might also want to check if the `segments` iterator is empty and return an error if so since we do not want to allow users to implement something for the limited Rust prelude anyway. We are already implementing the prelude.

Reply:
Since we want to motivate using the `module_path!` macro we may want to provide an additional `from_str` constructor that expects the input of the `module_path!` macro, e.g. some `&'static str` that is delimited by `::`: `foo::bar::baz`.

Reply:
Or we can also provide a macro that does the job for us.